llvm/lib/Target/RISCV/GISel/RISCVInstructionSelector.cpp

```diff
@@ -376,6 +391,120 @@ bool RISCVInstructionSelector::selectSExtInreg(MachineInstr &MI,
   return true;
 }
 
+/// Returns the RISCVCC::CondCode that corresponds to the CmpInst::Predicate CC.
+/// CC Must be an ICMP Predicate.
+static RISCVCC::CondCode getRISCVCCFromICMP(CmpInst::Predicate CC) {
+  switch (CC) {
+  default:
+    llvm_unreachable("Expected ICMP CmpInst::Predicate.");
+  case CmpInst::Predicate::ICMP_EQ:
+    return RISCVCC::COND_EQ;
+  case CmpInst::Predicate::ICMP_NE:
+    return RISCVCC::COND_NE;
+  case CmpInst::Predicate::ICMP_ULT:
+    return RISCVCC::COND_LTU;
+  case CmpInst::Predicate::ICMP_SLT:
+    return RISCVCC::COND_LT;
+  case CmpInst::Predicate::ICMP_UGE:
+    return RISCVCC::COND_GEU;
+  case CmpInst::Predicate::ICMP_SGE:
+    return RISCVCC::COND_GE;
+  }
+}
+
+void RISCVInstructionSelector::getICMPOperandsForBranch(
+    MachineInstr &MI, MachineIRBuilder &MIB, MachineRegisterInfo &MRI,
+    RISCVCC::CondCode &CC, Register &LHS, Register &RHS) const {
+  assert(MI.getOpcode() == TargetOpcode::G_ICMP);
+  CmpInst::Predicate ICMPCC =
+      static_cast<CmpInst::Predicate>(MI.getOperand(1).getPredicate());
+  LHS = MI.getOperand(2).getReg();
+  RHS = MI.getOperand(3).getReg();
+
+  // Adjust comparisons to use comparison with 0 if possible.
+  MachineInstr *MaybeConstant = MRI.getVRegDef(RHS);
+  if (MaybeConstant && MaybeConstant->getOpcode() == TargetOpcode::G_CONSTANT) {
```

Comment:
Maybe we should be using `getIConstantVRegValWithLookThrough`?

Reply:
I used `matchConstant` which uses the util function which gets the sign extended constant.

Reply:
The `getIConstantVRegValWithLookThrough` gives you more of the global aspect. The constant could sit in another block and the hit rate is higher. `MRI.getVRegDef(RHS)`could be a `COPY`.

Reply:
See https://github.com/llvm/llvm-project/blob/8902f12e61aa2d3053ad4d536e36569fc8bc9512/llvm/lib/Target/AArch64/GISel/AArch64InstructionSelector.cpp#L715

Reply:
It looks like the approach @tschuett may be better than using m_ICst that this patch is currently using, since its taking advantage of the look through. I have updated this patch to use `getIConstantVRegSExtVal` which calls `getIConstantVRegValWithLookThrough` but returns the sign extended constant, so that we look through COPYs.

Reply:
Sorry, but it does not look through:
https://github.com/llvm/llvm-project/blob/9f87522b121a08b97aff433e8b53dec1acf7c9d0/llvm/lib/CodeGen/GlobalISel/Utils.cpp#L295

Reply:
I created a patch https://github.com/llvm/llvm-project/pull/68327 that could be leveraged here to avoid the duplication of extracting the value from the optional and sign extending.

Reply:
Do tablegen patterns looking for constants, use look through? There is no reason to be stronger than tablegen here.

Reply:
I'm asking about GlobalISel's reuse of tblgen patterns.

Reply:
I am confused. Are you referring to:
https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Target/TargetSelectionDAG.td

GISEL uses tblgen for it's combiner.

Reply:
I'm referring to the instruction matching table in files like `RISCVGenGlobalISel.inc` which are created from the same patterns used for SelectionDAG.

For example here is one of the matches for an instruction with an immediate. It does not appear to do any look through.

```
        GIM_CheckOpcode, /*MI*/1, TargetOpcode::G_CONSTANT,                      
        GIM_CheckI64ImmPredicate, /*MI*/1, /*Predicate*/GICXXPred_I64_Predicate_simm12,
```

The code in this patch probably could have been written with a lot of tblgen patterns instead of custom C++ code. I don't see why we should use look through here if we wouldn't have gotten look through using a tblgen pattern.

Reply:
>  I don't see why we should use look through here if we wouldn't have gotten look through using a tblgen pattern.

Why's that? Wouldn't we prefer an implementation that is more capable of optimizing?

Reply:
The C++ code for the G_ICMP optimization with look through ought to better than the DAG code. I have the corresponding AArch64 file. It seems to be the import of the ISD::** for AArch64.

Reply:
> > I don't see why we should use look through here if we wouldn't have gotten look through using a tblgen pattern.
> 
> Why's that? Wouldn't we prefer an implementation that is more capable of optimizing?

We're still in the bring up phase focusing on -O0 not necessarily optimization. It feels premature to me to make a G_ICMP used by a select capable of being optimized differently than a G_ICMP by itself. Look through isn't free, it would have a compile time cost, though it is probably small it is somewhat counter to -O0. With optimizations enabled, shouldn't constant folding combines have already folded the constant chain? Are we talking about a real optimization opportunity or just something hypothetical? If there's a real missed opportunity than it feels like improving constant folding would give more improvement so that the imported tblgen patterns get the benefit since there are many more of those.

Reply:
Did we come to agreement on whether to use look through or not?

Reply:
I think we should commit the patch as is. I anticipate more refining of how we handle compares and selects/branches in the future to do optimizations.

Reply:
I will commit as is and refinement in the future can be made.